### PR TITLE
Adding `FormTextField` component

### LINKED ISSUE #16152: [Ext Nav] Create component: FormTextField
### Description
Create a reusable UI component for `FormTextField`

### References
[Figma Link](https://www.figma.com/file/HKpPKij9V3TpsyMV1TpV7C/DS-Components?node-id=7294%3A24987)
[UI component guidelines](https://www.notion.so/UI-Components-5ccd2bf83eb8441892a0b72c0d8929e1)
[Styling guidelines](https://www.notion.so/Extension-Frontend-Engineering-Guide-b82ddb3e14004b5db3799a9b446294a9#49aa2ef7b7aa40178ff918043df71c34)
Testing guidelines: TBC

### Files needed
- `index.js`
- `index.scss`
- `form-text-field.js`
- `form-text-field.stories.js`
- `form-text-field.test.js`
- `README.mdx`

### Technical details
TBC

### Acceptance criteria
- [ ] Uses color, typography, shadows design tokens 
- [ ] Uses semantic html
- [ ] PropTypes have descriptions
- [ ] Has storybook story with controls
- [ ] Has documentation in MDX
- [ ] Has unit tests and 90% coverage
- [ ] Works in Chrome and Firefox
- [ ] Performance tested: no unnecessary re-renders or other performance concerns

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation

Adding `FormTextField` that bundles `TextField`, `Label` and `HelpText` components together to create forms

* Fixes #16152 

### After

https://user-images.githubusercontent.com/8112138/201794011-01ddafda-8374-4261-b983-1da18e0bb301.mov

## Manual Testing Steps

- Go to the latest build of Storybook on this PR
- Search `FormTextField` in the search bar
- Click on the result
- Check story and controls
- Check MDX docs
- Check console for any errros

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/form-text-field/README.mdx [hunk 1]
```diff
@@ -0,0 +1,278 @@
+import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
+
+import { FormTextField } from './form-text-field';
+
+# FormTextField
+
+The `FormTextField` allows users to enter text to search. It wraps the `TextField` component that adds a search icon to the left of the input.
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/README.mdx:7]: Update to:
```suggestion
The `FormTextField` is an input component to create forms. It bundles the `Label, `TextField` and `HelptText` component together
```

File: ui/components/component-library/form-text-field/README.mdx [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+<Canvas>
+  <Story id="ui-components-component-library-form-text-field-form-text-field-stories-js--default-story" />
+</Canvas>
+
+## Props
+
+The `FormTextField` accepts all props below as well as all [Box](/docs/ui-components-ui-box-box-stories-js--default-story#props), [TextField](/docs/ui-components-component-library-text-field-text-field-stories-js--default-story#props) and [TextFieldBase](/docs/ui-components-component-library-text-field-base-text-field-base-stories-js--default-story#props) component props
+
+<ArgsTable of={FormTextField} />
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/README.mdx:17]: I guess I need to add all the different seeing as I am exposing them and passing them to each child component sigh.

- [ ] Add all missing propTypes

Also weird that these are showing up as `unknown`

Screen Shot 2022-11-14 at 6 14 19 PM

File: ui/components/component-library/form-text-field/form-text-field.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+   */
+  textFieldProps: PropTypes.object,
+  /**
+   * Props that are applied to the HelpText component
+   */
+  helpTextProps: PropTypes.object,
+  /**
+   * FormTextField accepts all the props from TextField and Box
+   */
+  ...TextField.propTypes,
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/form-text-field.js:142]: Add missing prop types `label`, `helpText` and check if there are any others

File: ui/components/component-library/form-text-field/form-text-field.scss [hunk 4]
```diff
@@ -0,0 +1,9 @@
+.mm-form-text-field {
+  --help-text-opacity-disabled: 0.5;
+
+  &--disabled {
+    .mm-form-text-field__help-text {
+      opacity: var(--help-text-opacity-disabled);
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/form-text-field.scss:6]: Not sure about adding opacity to the help text. @garrettbear or @brianacnguyen do you have any opinions on this? Could we ever want the help text to be normal opacity that describes why the input is disabled?

Normal

Screen Shot 2022-11-14 at 6 19 06 PM

Disabled

Screen Shot 2022-11-14 at 6 18 34 PM


cc @Akatori-Design

File: ui/components/component-library/form-text-field/form-text-field.test.js [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    return (
+      <FormComponent
+        value={value}
+        onChange={(e) => setValue(e.target.value)}
+        {...props}
+      />
+    );
+  };
+  return { user: userEvent.setup(), ...render(<ControlledWrapper />) };
+}
```
Comment (georgewrmarshall) [inline ui/components/component-library/form-text-field/form-text-field.test.js:33]: These should be moved to `test/lib/render-helpers.js` now that they are being used more than once.

File: ui/components/component-library/text-field-base/text-field-base.js [hunk 6]
```diff
@@ -155,7 +155,7 @@ TextFieldBase.propTypes = {
   /**
    * Autocomplete allows the browser to predict the value based on earlier typed values
    */
-  autoComplete: PropTypes.string,
+  autoComplete: PropTypes.bool,
   /**
    * If `true`, the input will be focused during the first mount.
    */
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.js:161]: Move this to it's own PR

File: ui/components/component-library/text-field-base/text-field-base.test.js [hunk 7]
```diff
@@ -236,7 +233,9 @@ describe('TextFieldBase', () => {
     );
   });
   it('should render with a custom input and still work', () => {
-    const CustomInputComponent = (props) => <input {...props} />;
+    const CustomInputComponent = React.forwardRef((props, ref) => (
+      <Box ref={ref} as="input" {...props} />
+    ));
     const { getByTestId } = render(
       <TextFieldBase
         InputComponent={CustomInputComponent}
```
Comment (georgewrmarshall) [inline ui/components/component-library/text-field-base/text-field-base.test.js:241]: Move to separate PR

File: ui/components/component-library/form-text-field/README.mdx (hunk 1 above)
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/README.mdx:7]: done

File: ui/components/component-library/text-field-base/text-field-base.test.js (hunk 7 above)
Comment (georgewrmarshall) [reply ui/components/component-library/text-field-base/text-field-base.test.js:241]: done https://github.com/MetaMask/metamask-extension/pull/16508

File: ui/components/component-library/text-field-base/text-field-base.js (hunk 6 above)
Comment (georgewrmarshall) [reply ui/components/component-library/text-field-base/text-field-base.js:161]: done https://github.com/MetaMask/metamask-extension/pull/16508

File: test/lib/render-helpers.js [hunk 8]
```diff
@@ -137,3 +137,12 @@ export function renderControlledInput(InputComponent, props) {
...
 }
+
+// userEvent setup function as per testing-library docs
+// https://testing-library.com/docs/user-event/intr
+export function renderWithUserEvent(jsx) {
+  return {
+    user: userEvent.setup(),
+    ...render(jsx),
+  };
+}
```
Comment (georgewrmarshall) [inline test/lib/render-helpers.js:148]: Moving to render-helpers to be used across form component tests

File: ui/components/component-library/form-text-field/form-text-field.scss (hunk 4 above)
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.scss:6]: OK thanks! Let's leave it as is for now 🙏

File: ui/components/component-library/form-text-field/README.mdx (hunk 2 above)
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/README.mdx:17]: I think we have a resolution for this now! Which is to include the children component ArgTables in the parent component MDX documentation!

File: ui/components/component-library/form-text-field/form-text-field.stories.js [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+
+export const DefaultStory = Template.bind({});
+DefaultStory.storyName = 'Default';
+
+export const id = Template.bind({});
+id.args = {
+  id: 'accessible-input-id',
+  label: 'If label prop exists id prop is required for accessibility',
+  helpText: '',
+};
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js:228]: Of course 🙏 Massive 🧠💨 on my part 🤦

File: ui/components/component-library/form-text-field/form-text-field.stories.js [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+  };
+  const VALIDATED_VALUES = {
+    NETWORK_NAME: 'network name',
+    NEW_RPC_URL: 'new rpc url',
+    CHAIN_ID: 'chain id',
+  };
+  const ERROR_MESSAGES = {
+    NETWORK_NAME: `Please enter enter "${VALIDATED_VALUES.NETWORK_NAME}"`,
+    NEW_RPC_URL: `Please enter enter "${VALIDATED_VALUES.NEW_RPC_URL}"`,
+    CHAIN_ID: `Please enter enter "${VALIDATED_VALUES.CHAIN_ID}"`,
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js:288]: Nice catch thanks!

File: ui/components/component-library/form-text-field/form-text-field.stories.js [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+          marginBottom={4}
+          label="Network name"
+          placeholder="Enter 'network name'"
+          required
+          name="networkName"
+          id="networkName"
+          onChange={handleOnChange}
+          value={values.networkName}
+          error={Boolean(submitted === FORM_STATE.ERROR && errors.networkName)}
+          helpText={submitted === FORM_STATE.ERROR && errors.networkName}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js:369]: Ahh yup that would return as a boolean if it's false haha. I think I should be able to get away with just `submitted === FORM_STATE.ERROR ? errors.networkName : null`

File: ui/components/component-library/form-text-field/form-text-field.stories.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js]: Updated https://github.com/MetaMask/metamask-extension/pull/16497/commits/acca6ba4a34d6bc73fe1e23b0fb7f1bbae32d98e

File: ui/components/component-library/form-text-field/form-text-field.stories.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js]: Updated https://github.com/MetaMask/metamask-extension/pull/16497/commits/acca6ba4a34d6bc73fe1e23b0fb7f1bbae32d98e

File: ui/components/component-library/form-text-field/form-text-field.stories.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js]: Updated https://github.com/MetaMask/metamask-extension/pull/16497/commits/acca6ba4a34d6bc73fe1e23b0fb7f1bbae32d98e

File: ui/components/component-library/form-text-field/form-text-field.stories.js [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+          marginBottom={4}
+          label="New RPC URL"
+          placeholder="Enter 'new RPC URL'"
+          required
+          name="newRpcUrl"
+          id="newRpcUrl"
+          onChange={handleOnChange}
+          value={values.newRpcUrl}
+          error={Boolean(submitted === FORM_STATE.ERROR && errors.newRpcUrl)}
+          helpText={submitted === FORM_STATE.ERROR && errors.newRpcUrl}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js]: Updated https://github.com/MetaMask/metamask-extension/pull/16497/commits/acca6ba4a34d6bc73fe1e23b0fb7f1bbae32d98e

File: ui/components/component-library/form-text-field/form-text-field.stories.js [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+          label="Chain ID"
+          marginBottom={4}
+          placeholder="Enter 'chain ID'"
+          required
+          name="chainId"
+          id="chainId"
+          onChange={handleOnChange}
+          value={values.chainId}
+          error={Boolean(submitted === FORM_STATE.ERROR && errors.chainId)}
+          helpText={submitted === FORM_STATE.ERROR && errors.chainId}
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/form-text-field/form-text-field.stories.js]: Updated https://github.com/MetaMask/metamask-extension/pull/16497/commits/acca6ba4a34d6bc73fe1e23b0fb7f1bbae32d98e

File: test/lib/render-helpers.js (hunk 8 above)
Comment (DimaJah) [reply test/lib/render-helpers.js:148]: []()~~~~

**Review by garrettbear — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]
